flink-connector-mongodb/src/test/java/org/apache/flink/connector/mongodb/table/MongoDynamicTableSourceITCase.java

```diff
@@ -0,0 +1,370 @@
+            List<String> expected =
+                    Arrays.asList(
+                            "+I[1, Alice, 1, 2, false]",
+                            "+I[1, Alice, 1, 2, false]",
+                            "+I[2, Bob, 2, 2, false]");
+
+            assertThat(result).hasSize(3);
+            assertThat(result).isEqualTo(expected);
+            if (caching == Caching.ENABLE_CACHE) {
+                // Validate cache
+                Map<String, LookupCacheManager.RefCountedCache> managedCaches =
+                        LookupCacheManager.getInstance().getManagedCaches();
+                assertThat(managedCaches).hasSize(1);
+                LookupCache cache =
+                        managedCaches.get(managedCaches.keySet().iterator().next()).getCache();
+                validateCachedValues(cache);
+            }
+
+        } finally {
+            if (caching == Caching.ENABLE_CACHE) {
+                LookupCacheManager.getInstance().checkAllReleased();
+                LookupCacheManager.getInstance().clear();
+                LookupCacheManager.keepCacheOnRelease(false);
+            }
+        }
+    }
+
+    private static void validateCachedValues(LookupCache cache) {
+        // mongo does support project push down, the cached row has been projected
+        RowData key1 = GenericRowData.of(1L);
+        RowData value1 = GenericRowData.of(1L, StringData.fromString("2"), false);
+
+        RowData key2 = GenericRowData.of(2L);
+        RowData value2 = GenericRowData.of(2L, StringData.fromString("2"), false);
+
+        RowData key3 = GenericRowData.of(3L);
+
+        Map<RowData, Collection<RowData>> expectedEntries = new HashMap<>();
+        expectedEntries.put(key1, Collections.singletonList(value1));
+        expectedEntries.put(key2, Collections.singletonList(value2));
+        expectedEntries.put(key3, Collections.emptyList());
+
+        LookupCacheAssert.assertThat(cache).containsExactlyEntriesOf(expectedEntries);
+    }
+
+    private enum Caching {
+        ENABLE_CACHE,
+        DISABLE_CACHE
+    }
+
+    private static String createTestDDl(Map<String, String> extraOptions) {
+        Map<String, String> options = new HashMap<>();
+        options.put(CONNECTOR.key(), "mongodb");
+        options.put(URI.key(), MONGO_CONTAINER.getConnectionString());
+        options.put(DATABASE.key(), TEST_DATABASE);
+        options.put(COLLECTION.key(), TEST_COLLECTION);
+        if (extraOptions != null) {
+            options.putAll(extraOptions);
+        }
+
+        String optionString =
+                options.entrySet().stream()
+                        .map(e -> String.format("'%s' = '%s'", e.getKey(), e.getValue()))
+                        .collect(Collectors.joining(",\n"));
+
+        return String.join(
+                "\n",
+                Arrays.asList(
+                        "CREATE TABLE mongo_source",
+                        "(",
+                        "  _id BIGINT,",
```

Comment:
Maybe I'm laking some MongoDB knowledge here but why do we allow `_id` to be BIGINT? Shouldn't this be always a hexadecimal string. I'm wondering whether we actually support round-trip semantics with this PR. A connector should feel like a database table which means the content of the sink can be re-read by the source ideally with exactly the same schema? Is this currently possible?

Reply:
The MongoDB's `_id` is similar to Elasticsearch's `_id`. 

The field name _id is reserved for use as a primary key;  Its value must be unique in the collection, is immutable, and may be of any type other than an array. 

https://www.mongodb.com/docs/manual/core/document/#the-_id-field
> In MongoDB, each document stored in a collection requires a unique [_id](https://www.mongodb.com/docs/manual/reference/glossary/#std-term-_id) field that acts as a [primary key](https://www.mongodb.com/docs/manual/reference/glossary/#std-term-primary-key). If an inserted document omits the _id field, the MongoDB driver automatically generates an [ObjectId](https://www.mongodb.com/docs/manual/reference/bson-types/#std-label-objectid) for the _id field.
>
> The _id field has the following behavior and constraints:
> - By default, MongoDB creates a unique index on the _id field during the creation of a collection.
> - The _id field may contain values of any [BSON data type](https://www.mongodb.com/docs/manual/reference/bson-types/), other than an array, regex, or undefined.
> - _ If the _id contains subfields, the subfield names cannot begin with a ($) symbol.

The round-trip semantics is already supported with this PR.
The content of the sink can be re-read by the source ideally with exactly the same schema.
However, since some mongodb-specific types cannot be directly mapped to Flink sql types, when converting `BsonSymbol`, `BsonRegularExpression`, `BsonJavaScript`, `BsonDbPointer` to String, we may lose the original type information.
In other words, reading from one collection of MongoDB and writing to another collection of MongoDB, the schema of the two collections of may be different. Suppose we map `BsonSymbol` to Flink sql `String`, when we write string to MongoDB again, we are not sure whether it is mapped to `BsonSymbol` or `BsonString`.

But if we convert mongodb-specific types into extend-json format, we can keep the original types, restore the original types accurately, and let users use UDF to process them.

https://www.mongodb.com/docs/manual/reference/mongodb-extended-json/

Reply:
Thanks for the detailed explanation. I thought `_id` always contains an `ObjectId` that might be generated by the user -provided `_id`. Your explanation makes sense to me. So only the the missing data types should limit a round-trip but this we can tackle in future versions.